numpy/core/src/common/simd/avx512/arithmetic.h

```diff
@@ -112,5 +112,61 @@ NPY_FINLINE __m512i npyv_mul_u8(__m512i a, __m512i b)
 // TODO: emulate integer division
 #define npyv_div_f32 _mm512_div_ps
 #define npyv_div_f64 _mm512_div_pd
+NPY_FINLINE float npyv__sum_f32(__m256 a)
+{
+    __m128 t1 = _mm_add_ps(_mm256_castps256_ps128(a), _mm256_extractf128_ps(a,1));
+    __m128 t2 = _mm_movehdup_ps(t1);
+    __m128 t3 = _mm_add_ps(t1, t2);
+    __m128 t4 = _mm_movehl_ps(t3, t3);
+    __m128 t5 = _mm_add_ss(t3, t4);
+    return _mm_cvtss_f32(t5);
+}
+
+NPY_FINLINE double npyv__sum_f64(__m256d a)
+{
+    __m128d t1 = _mm_add_pd(_mm256_castpd256_pd128(a), _mm256_extractf128_pd(a,1));
+    __m128d t2 = _mm_unpackhi_pd(t1, t1);
+    __m128d t3 = _mm_add_sd(t2, t1);
+    return _mm_cvtsd_f64(t3);
+}
+
+NPY_FINLINE __m256 get_low_f32(__m512 a)
+{
+    return _mm512_castps512_ps256(a);
+}
+
+NPY_FINLINE __m256 get_high_f32(__m512 a)
+{
+    return _mm256_castpd_ps(_mm512_extractf64x4_pd(_mm512_castps_pd(a),1));
```

Comment:
Why not `_mm512_extractf32x4_ps`?

Reply:
you are right, `_mm512_extractf32x8_ps` is better.